### PR TITLE
fix: Correctly generate nested merged dictionaries

### LINKED ISSUE #13724: Nested MergedDictionaries not being included in XAML code gen
### Current behavior

Reinitializing a ResourceDictionary (A) and attempting to override resources that are referenced in A by adding a second ResourceDictionary (B) to A's MergedDictionaries does not work

Given the following xaml:

```xml
<StackPanel HorizontalAlignment="Center"
			VerticalAlignment="Center"
			Spacing="20">

	<Button Background="{ThemeResource MyBrush}"
			Content="Should Be Red"
			Width="200"
			Height="200" />

	<Border>
		<Border.Resources>
			<SolidColorBrush x:Key="MyBrush" Color="Blue" />
		</Border.Resources>
		<Button Background="{ThemeResource MyBrush}"
				Width="200"
				Content="Should Be Blue"
				Height="200" />
	</Border>

	<Border>
		<Border.Resources>
			<ResourceDictionary>
				<ResourceDictionary.MergedDictionaries>
					<ResourceDictionary Source="ms-appx:///UnoApp67/Styles/MyBrush.xaml">
						<ResourceDictionary.MergedDictionaries>
							<ResourceDictionary Source="ms-appx:///UnoApp67/Styles/MyColorGreen.xaml" />
						</ResourceDictionary.MergedDictionaries>
					</ResourceDictionary>
				</ResourceDictionary.MergedDictionaries>
			</ResourceDictionary>
		</Border.Resources>
		<Button Background="{ThemeResource MyBrush}"
				Width="200"
				Content="Should Be Green"
				Height="200" />
	</Border>
</StackPanel>
```

Where `MyBrush` is originally setup to use Red as its color in the AppResources.xaml:

```xml
<ResourceDictionary.MergedDictionaries>
	<XamlControlsResources xmlns="using:Microsoft.UI.Xaml.Controls" />
	<ResourceDictionary Source="ms-appx:///UnoApp67/Styles/MyColorRed.xaml" />
	<ResourceDictionary Source="ms-appx:///UnoApp67/Styles/MyBrush.xaml" />
</ResourceDictionary.MergedDictionaries>
```

Where MyBrush.xaml contains:

```xml
<ResourceDictionary.ThemeDictionaries>
	<ResourceDictionary x:Key="Light">
		<SolidColorBrush x:Key="MyBrush" Color="{StaticResource MyColor}" />
	</ResourceDictionary>
	<ResourceDictionary x:Key="Default">
		<SolidColorBrush x:Key="MyBrush" Color="{StaticResource MyColor}" />
	</ResourceDictionary>
</ResourceDictionary.ThemeDictionaries>
````

Gives us the following mismatch of behavior compared to Windows:

Screenshot 2023-09-20 150451

### Expected behavior

We should see the same behavior as Windows where the third Button is also Green.

### How to reproduce it (as minimally and precisely as possible)

[UnoApp67.zip](https://github.com/unoplatform/uno/files/12676751/UnoApp67.zip)

### Workaround

_No response_

### Works on UWP/WinUI

Yes

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia, Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia

### NuGet package version(s)

Uno.WinUI 4.10.13

### Affected platforms

WebAssembly, Android, iOS, macOS (AppKit), Mac Catalyst, Skia (WPF), Skia (GTK on Linux/macOS/Windows), Skia (Linux Framebuffer)

### IDE

Visual Studio 2022

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): Fixes #13724

## PR Type

What kind of change does this PR introduce?

### <samp>🤖 Generated by Copilot at a26897a</samp>

This pull request adds support and tests for nested ResourceDictionaries with Source properties in XAML. It introduces a new extension method `AddMergedDictionaries` for the `ResourceDictionary` type, which is used by the XAML code generator to simplify the generated code. It also improves the error handling for nested ThemeDictionaries, which are not supported.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/SourceGenerators/Uno.UI.SourceGenerators.Tests/XamlCodeGeneratorTests/Out/TTLNMDWSRD/XamlCodeGenerator_MyResourceDictionary_92716e07ff456818f6d4125e055d4d57.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+						{
+							IsParsing = true,
+							MergedDictionaries = {
+							// Source not resolved statically, falling back on external resource retrieval.
+							global::Uno.UI.ResourceResolver.RetrieveDictionaryForSource("ms-appx:///Path/To/File1.xaml", "C:/Project/0/MyResourceDictionary.xaml")
+							.AddMergedDictionaries(
+							// MergedDictionaries
+							// Source not resolved statically, falling back on external resource retrieval.
+							global::Uno.UI.ResourceResolver.RetrieveDictionaryForSource("ms-appx:///Path/To/File2.xaml", "C:/Project/0/MyResourceDictionary.xaml")
+							)
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators.Tests/XamlCodeGeneratorTests/Out/TTLNMDWSRD/XamlCodeGenerator_MyResourceDictionary_92716e07ff456818f6d4125e055d4d57.cs:86]: Due to the syntactic position of the merged dictionary (being in collection initializer), it was hard to adjust for the nested dictionaries without the new `AddMergedDictionaries` extension method.

@kazo0 Let me know if this new code gen makes sense to you.

Comment (Youssef1313) [reply src/SourceGenerators/Uno.UI.SourceGenerators.Tests/XamlCodeGeneratorTests/Out/TTLNMDWSRD/XamlCodeGenerator_MyResourceDictionary_92716e07ff456818f6d4125e055d4d57.cs:86]: @jeromelaban I think we might want to consider moving away from collection initializers. While they are very nice syntax, they can sometimes complicate things for source generation and forces us to add new extension methods that modifies some instance state and then return the instance, and also having us to have flags in the source generator code to tell when we are in different syntactic contexts to use appropriate language punctuation for the specific context.

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs [hunk 2]
```diff
@@ -3033,6 +3045,23 @@ private void BuildDictionaryFromSource(IIndentedStringBuilder writer, XamlMember
...
+				{
+					writer.AppendLineInvariantIndented(".AddMergedDictionaries(");
+					BuildMergedDictionaries(writer, mergedDictionaries, isInInitializer: false, isSetExtensionMethod: true);
+					writer.AppendLineInvariantIndented(")");
+				}
+
+				var themeDictionaries = dictObject.Members.FirstOrDefault(m => m.Member.Name == "ThemeDictionaries");
+				if (themeDictionaries is not null)
+				{
+					writer.AppendLineIndented("#error Nested ThemeDictionaries are not yet supported.");
```
Comment (kazo0) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:3062]: Should we create an issue to track this?

**Review by kazo0 — COMMENTED** [review]
Looks like there's a bunch of generated cs files included, are those needed?

**Comment by jeromelaban** [comment]
> Looks like there's a bunch of generated cs files included, are those needed?

Generated files are expected, it's used to validate regressions in the generator's output.

File: src/Uno.UI/UI/Xaml/ResourceDictionaryExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace Uno.UI.Xaml;
+
+/// <summary>
+/// Helper for XAML code generation. Not intended to be used in apps outside of XAML generator.
+/// </summary>
+[EditorBrowsable(EditorBrowsableState.Never)]
+public static class ResourceDictionaryExtensions
+{
+	public static ResourceDictionary AddMergedDictionaries(this ResourceDictionary dictionary, params ResourceDictionary[] mergedDictionaries)
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/ResourceDictionaryExtensions.cs:14]: I wonder if those two should be placed in:
https://github.com/unoplatform/uno/blob/57d0fed00d8215014544557844e272bf2c375863/src/Uno.UI/Helpers/MarkupHelper.cs#L20

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs (hunk 2 above)
Comment (Youssef1313) [reply src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:3062]: https://github.com/unoplatform/uno/issues/13798

File: src/Uno.UI/UI/Xaml/ResourceDictionaryExtensions.cs (hunk 3 above)
Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/ResourceDictionaryExtensions.cs:14]: I'm fine either way. I'll move them.

**Review by jeromelaban — APPROVED** [review]

**Pull request merged**